### PR TITLE
Indents using path rules

### PR DESCRIPTION
**Pull request by guhanthuran**

### CODE REVIEW

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (jafeltra) [inline src/optimizer/plugins/AddPathRulesOptimizer.ts:19]: I think Mappings can have Path Rules on them as well, right? Should those be included here as well?

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (jafeltra) [inline test/optimizer/plugins/AddPathRulesOptimizer.test.ts:13]: These can be removed.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (jafeltra) [inline test/optimizer/plugins/AddPathRulesOptimizer.test.ts:44]: I'm not sure why this rule needs to be cloned. Can we just use `addressPathRule` directly in the expect statement?

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (jafeltra) [inline test/optimizer/plugins/AddPathRulesOptimizer.test.ts:88]: Similar to above, for this test and the ones below it, can we just use the rules directly and not clone them? For rules that don't change, it should still be okay to just use the rule directly.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (jafeltra) [inline test/optimizer/plugins/AddPathRulesOptimizer.test.ts:167]: It looks like this test is trying to test that path rules are included on Profiles, rather than Instances, which is good. The test description should include that, and this test needs some `expect` statements to actually test something.

File: test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts
Comment (jafeltra) [inline test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts:986]: It might be nice to include the `address.period`, `address.period.start`, and `address.period.end` paths that are tested in the `'should add path rules for deeper paths'` test (on lines 106-153) in order to ensure the deeper level paths in path rules are indented correctly as well.

**Review by jafeltra — CHANGES_REQUESTED** [review]
I really like the approach to this tricky task! I just have a few, pretty minor comments that should be addressed. I think it would be good to have either @cmoesel  or @mint-thompson review my comments and see if they agree. If they do agree with some of them, we can discuss who will make the updates!

File: src/exportable/ExportableInstance.ts
Comment (cmoesel) [reply src/exportable/ExportableInstance.ts:6]: Yeah, the fact that none of the SUSHI types accept `PathRule` made things tricky.  I think `@ts-ignore` is the way to go for now, but maybe we should consider updating SUSHI to allow `PathRule` on these. Since a `PathRule` is basically a no-op, I don't think it would do much harm to allow them and then ignore them if they _somehow_ ever made it into an instance in SUSHI.

Comment (jafeltra) [reply src/exportable/ExportableInstance.ts:6]: I don't think SUSHI even has a `PathRule` class to include. It's only in the grammar, so we would need to add a class that we don't really ever need, which feels a little weird.

Comment (jafeltra) [reply src/exportable/ExportableInstance.ts:6]: Lol true...

I don't love my options there and I don't know which one is weirder 😆

Comment (cmoesel) [reply src/exportable/ExportableInstance.ts:6]: I think I might actually have a use case for why we might _want_ to preserve `PathRule`s in SUSHI... I thought of it when reviewing the manual slice ordering PR in SUSHI.

Let's say that I have an array that is sliced -- but the slice requirements represent the whole value of the slice.  For example, in an `Observation` profile:
```
Profile: LabObservation
Parent: Observation
Description: "An Observation with category fixed to "
* category ^slicing.discriminator.type = #pattern
* category ^slicing.discriminator.path = "$this"
* category ^slicing.rules = #open
* category ^slicing.description = "Slice by category value"
* category ^slicing.ordered = false
* category contains lab 1..1
* category[lab] = $OBSCAT#laboratory
```

Now let's say I have an instance where I want the `laboratory` category to come first, and then a `blood-lab` category.  Theoretically, I should be able to do this:
```
* category[lab]
* category[+] = $CUSTOMSYSTEM#blood-lab
```

I don't need to assign `category[lab]` because the value is already fixed and will be populated for me. So I just use a `PathRule` to indicate the ordering I want.  Right now, I don't get what I want because the `PathRule` is stripped out in the importer.  But if it were preserved, then it would be considered when determining the ordering.

Similarly, if I have an optional slice with fixed values within it, maybe I want to indicate that the slice _should_ be exported in the instance by just including a `PathRule` to the slice. Right now, the `PathRule` would get stripped out in the exporter and the slice wouldn't be included at all.

So... maybe we should think more about supporting `PathRule`s all the way through the SUSHI process.

File: tsconfig.json
Comment (cmoesel) [reply tsconfig.json:27]: Agreed

File: src/exportable/ExportableInstance.ts
Comment (cmoesel) [reply src/exportable/ExportableInstance.ts:6]: So after further conversation, I think the approach is:
* Use `@ts-ignore` for now.
* Update SUSHI to support `PathRule` (because it's needed for other use cases there too).
* Then integrate it back into GoFSH, removing the `@ts-ignore`s.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts:12]: I assume you're talking about `ConstructInlineInstanceOptimizer`? If so, yeah, I think that makes sense because we'd also want those new inline instances to have proper indentation.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts:19]: I didn't think so, but you are exactly right:
```
mappingEntityRule:  mappingRule | insertRule | pathRule;
```
So yes, we should add that here I guess.  I also didn't realize we supported insert rules on mappings.  Go figure.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts:43]: Agreed.

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:68]: Yes, I know that and you know that and Guhan knew that, but... apparently `tsc` does not know that. When you delete that block, look what happens:
image

Comment (cmoesel) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:68]: I supposed we could add a comment saying we know that it's not necessary but it keeps `tsc` happy.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/AddPathRulesOptimizer.test.ts:13]: Yup.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (jafeltra) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts:12]: Oh yep, I was. I wrote this comment right after looking at that file and didn't write down the context my head was in.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (jafeltra) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts:19]: This is another fact I re-learned when I was also re-learning that SUSHI doesn't have a true PathRule class. Oops.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/AddPathRulesOptimizer.test.ts:44]: Agreed. There were some areas where cloning was helpful, but this is not one of them.

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (jafeltra) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:68]: Oh yeah sorry I think there was a typo in my comment. I was suggesting we either keep the `if` block and move it a little further up in the `forEach` loop (so it is under the other if block that exits early just to keep the logically similar blocks together), or we can add a line that defines that `rule = rule as ExportableAssignmentRule | ExportableCaretValueRule;` and get rid of the `if` block entirely. But the comment would also work and would clear up the confusion I had!

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/AddPathRulesOptimizer.test.ts:88]: Agreed. In some tests, the clone is helpful when we change the data in the expected rules -- but it doesn't seem we're doing that here.

File: test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts:986]: That would be fine. Since the `SimplifyRulePathContextsOptimizer` was not modified in this PR, the tests are intended just to confirm that this optimizer will correctly handle the outputs of the new `AddPathRulesOptimizer`.

File: tsconfig.json
Comment (cmoesel) [reply tsconfig.json:27]: In fact, I'm not really a fan of most of these changes (presumably made by prettier).  Personally, I like the `,` where it was before -- after the key-value, but before the comment. If you agree, I'd say to just revert this whole file.

File: test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts
Comment (jafeltra) [reply test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts:986]: I'm fine if we don't want to take this suggestion. I was just thinking it covered more of the types of things that are output by `AddPathRulesOptimizer`, but you're probably right that it isn't necessary.

File: tsconfig.json
Comment (jafeltra) [reply tsconfig.json:27]: I do think this was prettier. I'm good to revert the file.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [inline src/optimizer/plugins/AddPathRulesOptimizer.ts:37]: I don't understand why we have `parentList` and `seenPathList` -- because they seem to be used in similar ways a few lines down from here.

In fact, there is an issue w/ the current approach because only top-level paths are added to the `parentList` and only `slicedRulePaths` are added to the `seenPathList`.  That means if there is an assignment on a nested rule path (e.g., `a.b`) -- and then assignments on that path's children as well (`a.b.c`, `a.b.d`), then a path rule will be added for `a.b` because it's not in the `parentList` or the `seenPathList` (yet).

I have written two unit tests to show this.  One is on a top-level element, so it passes (thanks to `parentList`), but the other is on a nested path, so it fails.  I'll push those tests up after submitting this review.

I think the best approach would be to have a `seenPathList`, but _every_ rule path should be added to the list (included the path rule paths that are added in the loop). I don't think a separate `parentList` is necessary.

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:68]: Also note that (weirdly) `rulesToRemove.push(i)` also had to be moved from line 60 to line 69 for `tsc` to be happy.

**Review by cmoesel — CHANGES_REQUESTED** [review]
I've found some things that I think could be reworked to be simpler -- and one bug that needs to be fixed. I will push a broken test demonstrating the bug.

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:68]: Oddly enough, if I move the (technically unnecessary) `if` statement a few lines up, TSC becomes unhappy again.  And attempting to re-assign `rule` using a cast also does _not_ work.  I really don't know why TSC is so finicky about this, but the extra `if` statement _is_ necessary and it must be in this exact location!

Comment (jafeltra) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:68]: Huh. I really thought I had gotten this to work. I can try again when I re-review after your changes, but I'm sure you're right. That sounds very odd!

File: tsconfig.json
Comment (cmoesel) [reply tsconfig.json]: Reverted in edd6c61.

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:72]: Added a comment to explain it in b178575.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts:13]: Added an `after` clause in 112d742.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts]: Removed in e0dd13a.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts]: Added `mappings` in e0dd13a.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/AddPathRulesOptimizer.test.ts]: Tidied up the tests in 5df2936.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/AddPathRulesOptimizer.test.ts]: Tidied up the tests in 5df2936.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/AddPathRulesOptimizer.test.ts]: Tidied up the tests in 5df2936.

File: test/optimizer/plugins/AddPathRulesOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/AddPathRulesOptimizer.test.ts]: Tidied up the tests in 5df2936.

File: test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts:986]: Added this in 5df2936.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts]: Originally done in 4d2d472, but then made obsolete by the refactoring in cc1246c.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts]: Done in 4d2d472, but then reworked some more in the refactoring in cc1246c.

File: src/optimizer/plugins/AddPathRulesOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/AddPathRulesOptimizer.ts]: 4d2d472 removes `parentPath` and adds all paths to `seenPaths`.

**Comment by cmoesel** [comment]
I implemented the fixes for all issues raised, but then realized that there were still some use cases that wouldn't work correctly. For example, if two rules share a common ancestor, but it isn't the parent of both rules.  E.g.,
```
* address.country = "JP"
* address.period.start = "2021-01-01"
```
The previous approach missed this because it was only looking for common _direct parents_ on both rules. I added 4765374 to demonstrate the broken use cases. Then I updated the code in cc1246c to walk paths backwards looking for any common ancestor.

Ready for re-review, @jafeltra!

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (jafeltra) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts:72]: I am good with the comment, but I did look back to see what I had tried and I got something like this to work without TSC being unhappy:

```
resource.rules.forEach((r, i) => {
  if (!(r instanceof ruleType && getRulePath(r).startsWith(basePath))) {
    return;
  }
  let rule = r as ExportableAssignmentRule | ExportableCaretValueRule;

  rulesToRemove.push(i);
  // ...
```

I really don't mind the comment though, so I'm good with leaving this how you have it.

File: test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts
Comment (jafeltra) [reply test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts:986]: Do you think this first test is necessary to include separately from the test below it? I think the one below it is a superset of this test, and I'd be fine if we just want to keep the one.

**Review by jafeltra — COMMENTED** [review]
This looks good! I left a few comments/questions on the previous threads, but none of them were major. I did have one other question about a case that I'm not sure if we need to cover or not. 

If we have FSH rules that look like this:

```
Instance: ChainsawPatient
InstanceOf: Patient
Usage: #example
* address.city = "Tokyo"
* address.country = "JP"
* name.given = "Joe"
* address.postalCode = "01730"
* address.district = "13"
```

we would currently create a path rule for `address` that gets inserted before the first rule. Should it also create another path rule for `address` before `address.postalCode` for `postalCode` and `district`?

File: test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts
Comment (mint-thompson) [inline test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts:1055]: The path rule here would just be at `period`, since it's indented, right?

**Review by mint-thompson — CHANGES_REQUESTED** [review]
Other than a minor change to a comment, this looks to be all correct. I like the approach of making `AddPathRuleOptimizer` only responsible for adding path rules, as its name implies: the simplifying of rules is a separate job, and it keeps that implementation very simple.

File: test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts
Comment (cmoesel) [reply test/optimizer/plugins/SimplifyRulePathContextsOptimizer.test.ts:986]: Let's just keep it as-is so that if (some day) the shallow one works, but the deep one gets broken, it's more obvious.

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (cmoesel) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts]: Ah, you found the magic formula.  I like that better than the (well-documented) wonkiness I had before.  Fixed in [4e2a035](https://github.com/FHIR/GoFSH/pull/202/commits/4e2a035d296c7ebd3246a6ada77355f1f0123711).

**Comment by cmoesel** [comment]
@jafeltra asked:

> If we have FSH rules that look like this:
> 
> ```
> Instance: ChainsawPatient
> InstanceOf: Patient
> Usage: #example
> * address.city = "Tokyo"
> * address.country = "JP"
> * name.given = "Joe"
> * address.postalCode = "01730"
> * address.district = "13"
> ```
> 
> we would currently create a path rule for `address` that gets inserted before the first rule. Should it also create another path rule for `address` before `address.postalCode` for `postalCode` and `district`?

I thought about that, but then realized it's a very hypothetical scenario.  While a _human_ might do something silly like stick a `name` rule in the middle of `address` rules, GoFSH won't ever do that.  Since it is processing the JSON and spitting out the rules as it does so, the rules will naturally come out in fairly predictable orders.  Sure, we might not know if `name` comes _after_ or _before_ `address`, but we know that the `address` rules (or rules for any complex object) will always be grouped together. So I don't think we even need to worry about it!

**Comment by cmoesel** [comment]
@jafeltra and @mint-thompson -- ready for re-review!

File: src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts
Comment (jafeltra) [reply src/optimizer/plugins/ConstructInlineInstanceOptimizer.ts]: Yay glad it worked!

**Review by cmoesel — APPROVED** [review]

**Comment by cmoesel** [comment]
I'm going to do an interactive rebase on this branch so that we can more cleanly separate out Guhan's work and he gets some Git cred for it.  Otherwise it will all get squashed together and might give me the credit as the most recent committer on this.

**Comment by cmoesel** [comment]
@mint-thompson and @jafeltra -- my rebase has dismissed your stale reviews. Please re-approve.  Thanks!

**Review by mint-thompson — APPROVED** [review]

**Review by jafeltra — APPROVED** [review]

**Pull request merged**
